File: back-end/src/main/java/es/ull/project/application/usecase/vehicle/CreateVehicleUseCase.java
package es.ull.project.application.usecase.vehicle;

import es.ull.project.domain.entity.Vehicle;
import es.ull.project.domain.enumerate.VehicleType;
import es.ull.project.domain.valueobject.capacity.VehicleCapacityKilograms;
import es.ull.project.domain.valueobject.capacity.VehicleCapacityLiters;
import es.ull.project.domain.valueobject.cost.TransportationVariableCost;

/**
 * Use case for creating a new vehicle.
 */
public interface CreateVehicleUseCase {
    /**
     * Creates a new vehicle with the specified attributes.
     *
     * @param vehicleType        the type of vehicle
     * @param capacityKilograms the transport capacity in kilograms
     * @param CapacityLiters     the transport capacity in liters
     * @param costPerKilometer   the cost per kilometer for transportation
     * @return the created vehicle
     */
    Vehicle create(VehicleType vehicleType, VehicleCapacityKilograms capacityKilograms, VehicleCapacityLiters CapacityLiters, TransportationVariableCost costPerKilometer);
}

File: front-end/src/application/usecase/VehicleManagement/create-vehicle-use-case.ts
// CreateVehicleUseCase.ts
// Use case contract for creating a new vehicle

import type { Either, DataError } from '@ull-tfg/ull-tfg-typescript';
import { Vehicle } from '../../../domain/entity/vehicle';
import { VehicleType } from '../../../domain/enumerate/vehicle-type';
import { Capacity } from '../../../domain/valueobject/demand/capacity';
import { TransportationVariableCost } from '../../../domain/valueobject/cost/transportation-variable-cost';

/**
 * Use case for registering a new vehicle in the system.
 * Input: vehicleType, transportCapacity, costPerKilometer
 * Output: Vehicle (entity)
 */

// Command object for input data
export interface CreateVehicleCommand {
    vehicleType: VehicleType;
    transportCapacity: Capacity;
    costPerKilometer: TransportationVariableCost;
}

// Result type for the use case
export type CreateVehicleResult = Vehicle;

// Use case contract
export interface CreateVehicleUseCase {
    /**
     * Handles the creation of a new vehicle
     *
     * @param command Data needed to create the vehicle
     * @returns Either a DataError or the created Vehicle
     */
    execute(command: CreateVehicleCommand): Promise<Either<DataError, CreateVehicleResult>>;
}
